active/ItemsSourceView-Indexof/ItemsSourceView-IndexOf.md

```diff
@@ -41,6 +41,9 @@ If an element is not present in the collection, the method returns -1.
 
 # API Notes
 ItemsSourceView.IndexOf(IInspectable item): Returns index in the items source. If an item is not in the items source, it will return -1. Null is an acceptable value for IndexOf and calling IndexOf with null will return the index of null in the ItemsSource or -1 if the ItemsSource does not contain null.
+
+Note that to use `IndexOf` with pass by value types (e.g. int,float,etc) you need to create the ItemsSourceView with a collection that uses unboxing for values passed to `IndexOf`.
+Otherwise calling `IndexOf` will always return `-1` regardless of the items contained in the collection.
```

Comment:
I understand what this is about, but I don't understand the resolution. What does it mean that I need to use unboxing for values passed to IndexOf?

Reply:
Or can we remove this from the API doc because we're going to address it in the code?

Reply:
Oh, yes my wording is a bit difficult here and definitely needs improvement. So the idea is: primitive types get wrapped in objects (aka boxing). To be able to "find" them again, you need to get the primitive of those values again, that is compare by value and not by reference. So you need to unbox the value passed to IndexOf to be able to properly find a pass by value type in the collection.

If you have an idea how we could convey that in the most understandable and readable way, please let me know.

Reply:
I am not sure if we are gonna do that. Currently, we just don't deal with pass by value's. And in the case of any collection that has IndexOf, we don't have any control anyway.

Reply:
For the user of this API, I think all we can do is state a restriction that it doesn't work. But are we going to address it in the code instead?

Reply:
We could circumvent that for simple IIterable's by using our internal vector and go over the values manually instead of using the build in Vector.IndexOf function. But would it be much of a gain here?

Reply:
Note that `string` is a value type in C++/WinRT but a reference type in C# and C++/CX, so saying "pass by value types" is ambiguous.

I think this discussion in the documentation is going too deep and will end up confusing people. How about

> If the `ItemsSource` supports `IVector<Object>`, `IVectorView<Object>`, or `IObservableVector<Object>`, then the collection's `IndexOf` method is used.
>
> If the `ItemsSource` supports only `IIterable<Object>`, then the `IndexOf` method uses object identity. (Note that for strings and value types, object identity is performed on the boxed value.)

Because it's not the case that `IndexOf` will always return `-1` on value types. You could pass the same box back into `IndexOf` and it will find it.

Reply:
Great wording @oldnewthing ! I've merged it into the existing text.

Reply:
Should we also add IEnumerable to the list along with IIterable<Object> for C# developers ?

Reply:
Good question. Looking at the documentation, wouldn't IIterable be IEnumerable, because C# would use the projected type? 
I've updated the comment to note that IEnumerable is the appropiate .Net type for clarity.

Reply:
Yes they are. Since we have a lot of users in C#, calling out explicitly might make sense. @MikeHillberg @oldnewthing would you agree ?

Reply:
I think it's reasonable to mention the C# equivalents. Would need to wordsmith a bit to avoid separating the antecedent from consequent.

> The `IndexOf` method uses the `IndexOf` method of the `ItemsSource` if it supports `IVector<Object>`, `IVectorView<Object>`, or `IObservableVector<Object>` (in C#, `IList<object>`, `IReadOnlyList<object>`, or `IObservableVector<object>`).
>
> The `IndexOf` method uses object identity if the `ItemsSource` supports only `IIterable<Object>` (in C#, `IEnumerable<object>`). Note that for strings and value types, object identity is performed on the boxed value.

Reply:
Should we state "in C# ..." or "in .Net ..." for the differences? After all there are other .Net languages which could potentially use the classes.

Also, IReadOnlyList (the projection of IVectorView) does not provide IndexOf, so we need to leave that out unfortunately. Instead of having that long lists, would it be easier to have a document for C++ and one for .Net listing the interfaces where we would use IndexOf instead of implementing it ourself?

Something we could also do is simply say: In case of IIterable, we provide it, and in any interface that provides IndexOf, we use that. That way we wouldn't have to provide a list of interfaces that might get outdated

Reply:
I just left it at IVector for this spec, because I don't understand the editorial approach to this problem on docs.microsoft.com. When we copy this there we can get help on it from them.